This notebook is about me practicing Pandas on a real dataset: Covid 19 from the Kaggle website. 

We'll start by importing the modules we need:

In [1]:
import numpy as np
import pandas as pd

from sklearn.impute import SimpleImputer
df = pd.read_csv('F:\DATA Scientist\PACKT\covid_19_clean_complete\covid_19_clean_complete.csv')  # accessing and reading the dataset

df.head(15)  # showing the 1st 15 records only

Province/State       Country/Region       Lat        Long  \
0                            NaN          Afghanistan  33.93911   67.709953   
1                            NaN              Albania  41.15330   20.168300   
2                            NaN              Algeria  28.03390    1.659600   
3                            NaN              Andorra  42.50630    1.521800   
4                            NaN               Angola -11.20270   17.873900   
5                            NaN  Antigua and Barbuda  17.06080  -61.796400   
6                            NaN            Argentina -38.41610  -63.616700   
7                            NaN              Armenia  40.06910   45.038200   
8   Australian Capital Territory            Australia -35.47350  149.012400   
9                New South Wales            Australia -33.86880  151.209300   
10            Northern Territory            Australia -12.46340  130.845600   
11                    Queensland            Australia -27.46980  153.025100   
12               South Australia            Australia -34.92850  138.600700   
13                      Tasmania            Australia -42.88210  147.327200   
14                      Victoria            Australia -37.81360  144.963100   

          Date  Confirmed  Deaths  Recovered  Active             WHO Region  
0   2020-01-22          0       0          0       0  Eastern Mediterranean  
1   2020-01-22          0       0          0       0                 Europe  
2   2020-01-22          0       0          0       0                 Africa  
3   2020-01-22          0       0          0       0                 Europe  
4   2020-01-22          0       0          0       0                 Africa  
5   2020-01-22          0       0          0       0               Americas  
6   2020-01-22          0       0          0       0               Americas  
7   2020-01-22          0       0          0       0                 Europe  
8   2020-01-22          0       0          0       0        Western Pacific  
9   2020-01-22          0       0          0       0        Western Pacific  
10  2020-01-22          0       0          0       0        Western Pacific  
11  2020-01-22          0       0          0       0        Western Pacific  
12  2020-01-22          0       0          0       0        Western Pacific  
13  2020-01-22          0       0          0       0        Western Pacific  
14  2020-01-22          0       0          0       0        Western Pacific

## Modify the Data:

#### Removing Columns:

So now we'd like to remove columns that are not important to us, like the lat and long columns. 

In [2]:
df.drop(['Lat', 'Long'], 
       axis=1,  # drop only the column axis, not the row one
       inplace=True  # do not create a copy, apply these changes directly in the original array
       )
df.head()

Province/State Country/Region        Date  Confirmed  Deaths  Recovered  \
0            NaN    Afghanistan  2020-01-22          0       0          0   
1            NaN        Albania  2020-01-22          0       0          0   
2            NaN        Algeria  2020-01-22          0       0          0   
3            NaN        Andorra  2020-01-22          0       0          0   
4            NaN         Angola  2020-01-22          0       0          0   

   Active             WHO Region  
0       0  Eastern Mediterranean  
1       0                 Europe  
2       0                 Africa  
3       0                 Europe  
4       0                 Africa

#### Renaming Titles:

In [3]:
df.rename(columns={
    'Province/State': 'Province',
    'Country/Region': 'Country', 
    'WHO Region': 'Region'}, 
          inplace=True)
df.head()

Province      Country        Date  Confirmed  Deaths  Recovered  Active  \
0      NaN  Afghanistan  2020-01-22          0       0          0       0   
1      NaN      Albania  2020-01-22          0       0          0       0   
2      NaN      Algeria  2020-01-22          0       0          0       0   
3      NaN      Andorra  2020-01-22          0       0          0       0   
4      NaN       Angola  2020-01-22          0       0          0       0   

                  Region  
0  Eastern Mediterranean  
1                 Europe  
2                 Africa  
3                 Europe  
4                 Africa

#### Date Format Conversion:

The date format that we see in the dataset seems to be formatted properly but just to be sure we will format it anyway.

In [4]:
df['Date'] = pd.to_datetime(df['Date'])
df.head()

Province      Country       Date  Confirmed  Deaths  Recovered  Active  \
0      NaN  Afghanistan 2020-01-22          0       0          0       0   
1      NaN      Albania 2020-01-22          0       0          0       0   
2      NaN      Algeria 2020-01-22          0       0          0       0   
3      NaN      Andorra 2020-01-22          0       0          0       0   
4      NaN       Angola 2020-01-22          0       0          0       0   

                  Region  
0  Eastern Mediterranean  
1                 Europe  
2                 Africa  
3                 Europe  
4                 Africa

## Basic Info on the Dataset:

### Number of NaN entries?: 

***


NaN entries (Not A Number) are what we see in the Province column, they just mean that there are no Province name for that Country and/or Region.  

We need to know how many there are in the set. We'll use the `.info()` method for that. 

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49068 entries, 0 to 49067
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Province   14664 non-null  object        
 1   Country    49068 non-null  object        
 2   Date       49068 non-null  datetime64[ns]
 3   Confirmed  49068 non-null  int64         
 4   Deaths     49068 non-null  int64         
 5   Recovered  49068 non-null  int64         
 6   Active     49068 non-null  int64         
 7   Region     49068 non-null  object        
dtypes: datetime64[ns](1), int64(4), object(3)
memory usage: 3.0+ MB


From this output we clearly see the total number of entries: 49068.  
Among these we have one column that does not fill all the 49068 entries, it only has 14664.  
This means that the difference is the actual number of NaN entries in this column.

#### Removing NaN Entries:

We need to deal with these missing values NaN because they should at least return something that will be considered an entry, not a NaN. This way the dataset will be complete.  

Let's replace all the NaN with the `.fillna()` method:

In [6]:
df.fillna(
    'NA',
    inplace=True
)
df.head()

Province      Country       Date  Confirmed  Deaths  Recovered  Active  \
0       NA  Afghanistan 2020-01-22          0       0          0       0   
1       NA      Albania 2020-01-22          0       0          0       0   
2       NA      Algeria 2020-01-22          0       0          0       0   
3       NA      Andorra 2020-01-22          0       0          0       0   
4       NA       Angola 2020-01-22          0       0          0       0   

                  Region  
0  Eastern Mediterranean  
1                 Europe  
2                 Africa  
3                 Europe  
4                 Africa

Now let's check if we filled all these NaN entries:

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49068 entries, 0 to 49067
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Province   49068 non-null  object        
 1   Country    49068 non-null  object        
 2   Date       49068 non-null  datetime64[ns]
 3   Confirmed  49068 non-null  int64         
 4   Deaths     49068 non-null  int64         
 5   Recovered  49068 non-null  int64         
 6   Active     49068 non-null  int64         
 7   Region     49068 non-null  object        
dtypes: datetime64[ns](1), int64(4), object(3)
memory usage: 3.0+ MB


Indeed we did. The dataset is complete. 

## Accessing the Data:  

***


### Customizing the Table:

So our dataset is complete and we can now start to focus on parts of it.  

Let's say I am now only interested in a country focused dataset, meaning I want specific column entries that are all ordered according to the full list of Countries.  

The `.groupby()` method is here to do all these in one line of code:

In [10]:
countryfocus_df = df.groupby('Country')[  # Country is the column we want a focus on
[  # The following columns are those we want entries from
    'Confirmed',
    'Deaths',
    'Recovered'
]
].sum().reset_index()  # .sum and .reset_index neatly rebuild the dataframe 

countryfocus_df

Country  Confirmed  Deaths  Recovered
0           Afghanistan    1936390   49098     798240
1               Albania     196702    5708     118877
2               Algeria    1179755   77972     755897
3               Andorra      94404    5423      69074
4                Angola      22662    1078       6573
..                  ...        ...     ...        ...
182  West Bank and Gaza     233461    1370      61124
183      Western Sahara        901      63        648
184               Yemen      67180   17707      23779
185              Zambia     129421    2643      83611
186            Zimbabwe      50794     881      12207

[187 rows x 4 columns]

Great. Now let's add the date column:

In [12]:
countryfocus_df = df.groupby(['Country', 'Date'])[  # Note the [] that were added
[
    'Confirmed',
    'Deaths',
    'Recovered'
]
].sum().reset_index()

countryfocus_df

Country       Date  Confirmed  Deaths  Recovered
0      Afghanistan 2020-01-22          0       0          0
1      Afghanistan 2020-01-23          0       0          0
2      Afghanistan 2020-01-24          0       0          0
3      Afghanistan 2020-01-25          0       0          0
4      Afghanistan 2020-01-26          0       0          0
...            ...        ...        ...     ...        ...
35151     Zimbabwe 2020-07-23       2124      28        510
35152     Zimbabwe 2020-07-24       2296      32        514
35153     Zimbabwe 2020-07-25       2434      34        518
35154     Zimbabwe 2020-07-26       2512      34        518
35155     Zimbabwe 2020-07-27       2704      36        542

[35156 rows x 5 columns]

All Right. So now let's dig even deeper into the data. I want to see all the confirmed entries with a number greater than 100.  
We can get this with a simple conditional statement '>':

In [13]:
confirmedcases_focus_df = countryfocus_df[countryfocus_df[
'Confirmed']>100]

confirmedcases_focus_df

Country       Date  Confirmed  Deaths  Recovered
65     Afghanistan 2020-03-27        110       4          2
66     Afghanistan 2020-03-28        110       4          2
67     Afghanistan 2020-03-29        120       4          2
68     Afghanistan 2020-03-30        170       4          2
69     Afghanistan 2020-03-31        174       4          5
...            ...        ...        ...     ...        ...
35151     Zimbabwe 2020-07-23       2124      28        510
35152     Zimbabwe 2020-07-24       2296      32        514
35153     Zimbabwe 2020-07-25       2434      34        518
35154     Zimbabwe 2020-07-26       2512      34        518
35155     Zimbabwe 2020-07-27       2704      36        542

[19539 rows x 5 columns]

At the bottom of each returned table we get the total number of rows returned: 35156 and 19539 for the confirmed cases focus table.  
We can already deduct very simple observations:  
- there are about 39000 date entries 
- and among them about 19000 were days where over 100 cases were confirmed. 